ethereum/eip712/encoding.go

```diff
@@ -0,0 +1,214 @@
+package eip712
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/simapp/params"
+	"github.com/cosmos/cosmos-sdk/x/auth/migrations/legacytx"
+
+	cosmosTypes "github.com/cosmos/cosmos-sdk/types"
+	txTypes "github.com/cosmos/cosmos-sdk/types/tx"
+
+	apitypes "github.com/ethereum/go-ethereum/signer/core/apitypes"
+	ethermint "github.com/evmos/ethermint/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+)
+
+var (
+	ethermintProtoCodec codec.ProtoCodecMarshaler
+	ethermintAminoCodec *codec.LegacyAmino
+)
+
+// The process of unmarshaling SignDoc bytes into a SignDoc object requires having a codec
+// populated with all relevant message types. As a result, we must call this method on app
+// initialization with the app's encoding config.
+func SetEncodingConfig(cfg params.EncodingConfig) {
+	ethermintAminoCodec = cfg.Amino
+	ethermintProtoCodec = codec.NewProtoCodec(cfg.InterfaceRegistry)
+}
+
+// Get the EIP-712 object hash for the given SignDoc bytes by first decoding the bytes into
+// an EIP-712 object, then hashing the EIP-712 object to create the bytes to be signed.
+// See https://eips.ethereum.org/EIPS/eip-712 for more.
+func GetEIP712HashForMsg(signDocBytes []byte) ([]byte, error) {
+	typedData, err := GetEIP712TypedDataForMsg(signDocBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	domainSeparator, err := typedData.HashStruct("EIP712Domain", typedData.Domain.Map())
+	if err != nil {
+		return nil, fmt.Errorf("could not hash EIP-712 domain: %w", err)
+	}
+
+	typedDataHash, err := typedData.HashStruct(typedData.PrimaryType, typedData.Message)
+	if err != nil {
+		return nil, fmt.Errorf("could not hash EIP-712 primary type: %w", err)
+	}
+	rawData := []byte(fmt.Sprintf("\x19\x01%s%s", string(domainSeparator), string(typedDataHash)))
+
+	return rawData, nil
+}
+
+// Get the EIP-712 TypedData representation of the sign doc bytes for either Amino or
+// Protobuf encodings.
+func GetEIP712TypedDataForMsg(signDocBytes []byte) (apitypes.TypedData, error) {
```

Comment:
nit code style (godoc): always start the comments of public funcs with the name
```suggestion
// GetEIP712TypedDataForMsg returns the EIP-712 TypedData representation for either
// Amino or Protobuf encoded signature doc bytes.
func GetEIP712TypedDataForMsg(signDocBytes []byte) (apitypes.TypedData, error) {
```